### ISSUE TITLE
How to put Gluon-TS model is in production

### ISSUE THREAD

**Issue by islamalh** [awslabs/gluon-ts#869]
I have trained several model and saved them. I can load the model and by using `predict` method I can use it for perdition. In order to use it as service I am using waitress and some custom code. The result is pretty good but I need to send lot's of query which making the prediction quite slow. I have about 100 customer and I am predicting one month in future with hourly grain so I need to send 100x24x30 request which is taking few minutes. Is there a alternate way I can use the Gluon-Ts model in production environment instead of Waitress?

Thanks

**Comment by jaheba** [comment]
Hello @islamalh,

do you use SageMaker for training models?

> I have about 100 customer and I am predicting one month in future with hourly grain so I need to send 100x24x30 request which is taking few minutes.

This doesn't sound too bad to me. Can you describe your requirements?

**Comment by islamalh** [comment]
Hi @jaheba 
Unfortunately no I am not allowed to use Sagemaker but I did use Gluon-TS library for Training. (Initially I started with Sagemaker but we wanted not to use it). For Prediction I am loading the model and serving it besides waitress. Here is the server looks like 

```
class PredictionServer(object):
    app = None
    def __init__(self, model_path):
        self.app = Flask(__name__)
        self.predictor = self.load_model(model_path)
        self.app.add_url_rule('/model/account/predict', methods=['POST'], view_func=self.request_to_model)

    def request_to_model(self):
        inference_data = request.get_json(force=True)
        # inference_data = json.loads(from_server)

        data = {}

        keys = inference_data.keys()
        for key in keys:
            if key not in ['start','feat_dynamic_real','target','source', 'feat_static_cat']:
                return

        data['start'] =pd.Timestamp(inference_data['start'], freq="H")
        data['feat_dynamic_real'] = inference_data['feat_dynamic_real']
        
        data['target'] = inference_data['target']

        assert len(data['feat_dynamic_real']) ==23
        assert len(data['feat_dynamic_real'][0]) - len(data['target']) == 24

        response_forecast: SampleForecast = next( self.predictor.predict([data]))
        response = {
        'high' : [ i if i>0 else 0.0 for i in response_forecast.quantile(0.95).tolist()],
        'low'  : [ i if i>0 else 0.0 for i in response_forecast.quantile(0.10).tolist()],
        'pred' : [ i if i>0 else 0.0 for i in response_forecast.quantile(0.50).tolist()],
        }
        return json.dumps(response)


    def load_model(self, model_path):


        try:
            predictor = Predictor.deserialize(Path(model_path))
            print('Model loaded from %s'%model_path)
        except:
            print('Unable to load the model')
            sys.exit(1)
        return predictor
```

As you can see the requirement is to use predict `High`, `mid` and `low` instead of full distribution and The model is use to predict 24 hours ahead at a time and use today's predicted result as a history for tomorrow. This process for each account I am running for full month at the beginning of the month. The problem I am having right now is the waitress server is too slow. It's taking about ~15 minutes to perform prediction for this 720 hours(if the month is 30 days) x 100 (account) which is problematic.

**Comment by jaheba** [comment]
What algorithm are you using? I assume it is DeepAR?

Doing a quick estimation, you do 100 * 30 requests, which in 15 minutes, are ~0.3 seconds per requests. This sounds about right.

Have you thought about parallelising the requests? If you have multiple workers, you should be able to get faster results.

**Comment by islamalh** [comment]
Yes I did throw 20 pods (In Kubernetes) in the production but still it's taking ~9 minutes or so

**Comment by jaheba** [comment]
That's strange, I would assume a bigger steedup. Something you could try is to send multiple requests at once, thus saving a bit on overhead. However, I don't expect too much of an improvement.

Do the 9 minutes include startup time?

We've used batch-transform jobs in SageMaker, and are essentially seeing linear performance improvements when adding more instances to it.

**Comment by islamalh** [comment]
Yes I was able to get best 4.5 minutes but I am giving you average time. It does include startup time. I can send multiple request for different account but for same account I can't as next day prediction depends on previous day's forecast. (day-1) is a attribute. 

Is it possible to use `multi-model-server` from AWS for this?
